### PR TITLE
JAMES-3378 Filter keywords

### PR DESCRIPTION
**Pull request by remk**

### CODE REVIEW

**Comment by chibenwa** [comment]
test this please

Deployment tests

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 1]
```diff
@@ -858,6 +874,208 @@ trait EmailQueryMethodContract {
...
+        .extract
+        .body
+        .asString
+
+      assertThatJson(response)
+        .inPath("$.methodResponses[0][1].ids")
+        .isEqualTo(s"""["${messageWithoudFlagId.serialize()}"]""")
+    }
+  }
+
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1078]: - Can I do a search with `\Recent` or `\Deleted` flags? 

- Can I search an empty string, or a falg with invalid char, or of more than 256 chars?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala [hunk 2]
```diff
@@ -20,13 +20,16 @@
...
 import org.apache.james.mailbox.model.{MailboxId, MessageId}
 
 case class FilterCondition(inMailbox: Option[MailboxId],
                            inMailboxOtherThan: Option[Seq[MailboxId]],
                            before: Option[UTCDate],
-                           after: Option[UTCDate])
+                           after: Option[UTCDate],
+                           hasKeyword: Option[Keyword],
+                           notKeyword: Option[Keyword]
+                          )
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailQuery.scala:32]: That line break is not necessary?

**Review by chibenwa — COMMENTED** [review]
Good work!

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/query/CriterionConverter.java [hunk 3]
```diff
@@ -310,7 +310,7 @@ private QueryBuilder convertDateOperator(String field, SearchQuery.DateComparato
         case BEFORE:
             return rangeQuery(field).lte(upDateString);
         case AFTER:
-            return rangeQuery(field).gte(lowDateString);
+            return rangeQuery(field).gt(lowDateString);
```
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/query/CriterionConverter.java:313]: No tests ?

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/utils/search/MailboxFilter.scala [hunk 4]
```diff
@@ -49,4 +53,57 @@ object MailboxFilter {
...
+      List(ReceivedBefore, ReceivedAfter, HasKeyWord, NotKeyWord).foldLeft(new SearchQuery.Builder())((builder, filter) => filter.toQuery(builder, request))
+    }
+  }
+
+  case object ReceivedBefore extends QueryFilter {
+    override def toQuery(builder: SearchQuery.Builder, request: EmailQueryRequest): SearchQuery.Builder =  request.filter.flatMap(_.before) match {
+      case Some(before) =>
+        val strictlyBefore = new InternalDateCriterion(new DateOperator(DateComparator.BEFORE, Date.from(before.asUTC.toInstant), DateResolution.Second))
+        val sameDate = new InternalDateCriterion(new DateOperator(DateComparator.ON, Date.from(before.asUTC.toInstant), DateResolution.Second))
+        new SearchQuery.Builder()
```
Comment (chibenwa) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/utils/search/MailboxFilter.scala:73]: You need to carry other the existing builder overwize previous criteria are lost?

(here and in other QueryFilter?)

**Comment by chibenwa** [comment]
Force pushed to solve conflict

**Review by remk — COMMENTED** [review]
:+1: for the fixups

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/utils/search/MailboxFilter.scala [hunk 5]
```diff
@@ -79,7 +79,7 @@ object MailboxFilter {
     override def toQuery(builder: SearchQuery.Builder, request: EmailQueryRequest): SearchQuery.Builder =  request.filter.flatMap(_.after) match {
       case Some(after) => {
         val strictlyAfter = new InternalDateCriterion(new DateOperator(DateComparator.AFTER, Date.from(after.asUTC.toInstant), DateResolution.Second))
-        new SearchQuery.Builder()
+        builder
```
Comment (remk) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/utils/search/MailboxFilter.scala:82]: :+1:

**Review by chibenwa — APPROVED** [review]

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 6]
```diff
@@ -975,6 +975,107 @@ trait EmailQueryMethodContract {
...
+         |    "c1"]]
+         |}""".stripMargin
+
+    awaitAtMostTenSeconds.untilAsserted { () =>
+      val response = `given`
+        .header(ACCEPT.toString, ACCEPT_RFC8621_VERSION_HEADER)
+        .body(request)
+        .when
+        .post
+        .`then`
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1300]: Indent

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala [hunk 7]
```diff
@@ -975,6 +975,107 @@ trait EmailQueryMethodContract {
...
```
Comment (LanKhuat) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/EmailQueryMethodContract.scala:1250]: Indent

**Review by LanKhuat — APPROVED** [review]

**Review by remk — COMMENTED** [review]
:+1:

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**